tests/gen_code.c

```diff
@@ -0,0 +1,285 @@
+			    "####################################################################\n");
+		fprintf(fd, "# Execute with python : 'python3 %s'\n", gen_file);
+		fprintf(fd, "####################################################################\n");
+		fprintf(fd, "import sys\n");
+		fprintf(fd, "try:\n"
+			    "    import iio\n"
+			    "except:\n"
+			    "    # By default the iio python bindings are not in path\n"
+			    "    print('you must fix your PYTHONPATH to include iio')\n"
+			    "    exit()\n\n");
+		fprintf(fd, "def main():\n");
+	}
+}
+
+void gen_context (const char *uri_in)
+{
+	if (!fd)
+		return;
+
+	uri = strdup(uri_in);
+	if (lang == C_LANG) {
+		fprintf(fd, "\t/* Create IIO Context */\n"
+		    "\tIIO_ASSERT(ctx = iio_create_context_from_uri(\"%s\"));\n\n", uri);
+	} else if (lang == PYTHON_LANG) {
+		fprintf(fd, "    # Create IIO Context\n"
+			    "    try:\n"
+			    "        ctx = iio.Context(\"%s\")\n", uri);
+		fprintf(fd, "    except OSError as e:\n"
+			    "        print('Unable to open context %s')\n", uri);
+		fprintf(fd, "        texit()\n\n");
+	}
+}
+
+void gen_context_destroy()
+{
+	if (!fd)
+		return;
+
+	if (lang == C_LANG) {
+		fprintf(fd, "\n\t/* Close context at %s, can release/destroy things */\n", uri);
+		fprintf(fd, "\tiio_context_destroy(ctx);\n\treturn EXIT_SUCCESS;\n}\n");
+	} else if (lang == PYTHON_LANG) {
+		/* No need to close context in python
+		 * destroy is called when the destructor is called for iio objects
+		 */
+		fprintf(fd, "\nif __name__ == '__main__':\n"
+			    "    main()\n");
+	}
+
+	fclose(fd);
+	free(uri);
+}
+
+void gen_context_attr(const char *key)
+{
+	if (!fd)
+		return;
+
+	if (lang == C_LANG) {
+		fprintf(fd, "\t/* Read IIO Context attribute and return result as string */\n");
+		fprintf(fd, "\tval_str = iio_context_get_attr_value(ctx, \"%s\");\n", key);
+		fprintf(fd, "\tprintf(\"%s : %%s\\n\", val_str);\n", key);
+	} else if (lang == PYTHON_LANG) {
+		fprintf(fd, "    # Read IIO Context attribute and return result as string\n");
+		fprintf(fd, "    print('%s : ' + ctx.attrs[\"%s\"])\n", key, key);
+	}
+}
+
+void gen_dev(const struct iio_device *dev)
+{
+	if (!fd)
+		return;
+	if (lang == C_LANG) {
+		fprintf(fd, "\t/* Find IIO device in current context */\n");
+		fprintf(fd, "\tIIO_ASSERT(dev = iio_context_find_device(ctx, \"%s\"));\n\n",iio_device_get_name(dev));
+	} else if (lang == PYTHON_LANG) {
+		fprintf(fd, "    # Find IIO device in current context\n");
+		fprintf(fd, "    dev = ctx.find_device(\"%s\")\n\n",iio_device_get_name(dev));
+	}
+}
+
+void gen_ch(const struct iio_channel *ch)
+{
+	if (!fd)
+		return;
+
+	if (lang == C_LANG) {
+		fprintf(fd, "\t/* Find the IIO %s channel in the current device */\n",
+			iio_channel_is_output(ch) ? "Output" : "Input");
+		fprintf(fd, "\tIIO_ASSERT(ch = iio_device_find_channel(dev, \"%s\", %s));\n\n",
+			iio_channel_get_name(ch), iio_channel_is_output(ch) ? "true" : "false");
```

Comment:
Might want to use iio_channel_get_id, not sure if all channels have names.

Reply:
I think this does the same as iio_attr; the pattern matching happens based on get_name, and it prints out _id...

Reply:
I get these weird nulls sometimes with get_name
```
    # Find IIO device in current context
    dev = ctx.find_device("ad9361-phy")

    # Find the IIO Output channel in the current device
    ch = dev.find_channel("(null)", True)
    # Write string to channel attribute:
    ch.attrs["rf_port_select"].value = str("A")
    print("wrote A into rf_port_select")
```
Command: `iio_attr -o -u ip:analog -q -g /tmp/foo.py -c ad9361-phy voltage0 rf_port_select 'A'`

Reply:
OK - I will have a look - thanks for the test case.

Reply:
I think this is fixed in the latest force push